services/agent/src/simcore_service_agent/modules/task_monitor.py

```diff
@@ -0,0 +1,140 @@
+            except Exception as e:  # pylint: disable=broad-except
+                logger.error(
+                    "Had an error while running '%s':\n%s",
+                    coroutine.__name__,
+                    "\n".join(traceback.format_tb(e.__traceback__)),
+                )
+
+            if task_data.repeat_interval_s is None:
+                logger.warning(
+                    "Unexpected termination of '%s'; it will be restarted",
+                    coroutine.__name__,
+                )
+
+            logger.info(
+                "Will run '%s' again in %s seconds",
+                coroutine.__name__,
+                task_data.repeat_interval_s,
+            )
+            await asyncio.sleep(
+                DEFAULT_TASK_WAIT_ON_ERROR
+                if task_data.repeat_interval_s is None
+                else task_data.repeat_interval_s
+            )
+
+
+class TaskMonitor:
+    def __init__(self):
+        self._running: bool = False
+        self._tasks: set[asyncio.Task] = set()
+
+        self._to_start: dict[str, _TaskData] = {}
+
+    @property
+    def is_running(self) -> bool:
+        return self._running
+
+    def register_job(
+        self,
+        target: Callable,
+        *args: Any,
+        repeat_interval_s: Optional[PositiveFloat] = None,
+    ) -> None:
+        if self._running:
+            raise RecursionError(
+                "Cannot add more tasks, monitor already running with: "
+                f"{[x.get_name() for x in self._tasks]}"
+            )
+
+        if target.__name__ in self._to_start:
+            raise RuntimeError(f"{target.__name__} is already registered")
+
+        self._to_start[target.__name__] = _TaskData(target, args, repeat_interval_s)
+
+    async def start(self) -> None:
+        self._running = True
+        for name, task_data in self._to_start.items():
+            logger.info("Starting task '%s'", name)
+            self._tasks.add(
+                asyncio.create_task(_task_runner(task_data), name=f"task_{name}")
+            )
+
+    async def shutdown(self):
+        async def _wait_for_task(task: asyncio.Task) -> None:
+            with suppress(asyncio.CancelledError):
+                await task
+
+        tasks_to_wait: deque[Awaitable] = deque()
+        for task in set(self._tasks):
+            logger.info("Cancel and stop task '%s'", task.get_name())
+
+            task.cancel()
+            tasks_to_wait.append(_wait_for_task(task))
+            self._tasks.remove(task)
+
+        await asyncio.gather(*tasks_to_wait, return_exceptions=True)
+        self._running = False
+        self._to_start = {}
+
+
+def setup(app: FastAPI) -> None:
+    async def _on_startup() -> None:
+        task_monitor = app.state.task_monitor = TaskMonitor()
+        settings: ApplicationSettings = app.state.settings
+
+        # setup all relative jobs
+        task_monitor.register_job(
+            backup_and_remove_volumes,
+            settings,
```

Comment:
THOUGHT: why would the API of ``task_monitor.register_job``  is designed to get ``repeat_interval_s=`` from settings and you pass again all the settings inside? IMO this is a bad interface (we do that all the time) because on the one hand we make explicit that we need some settings (i.e. ``repeat_interval_s``) and at the same time we pass them all (i.e. ``settings``). I think we should have interfaces with one or the other option but not both!

Reply:
I think you are mixing up `task_monitor` logic with user defined handler logic. Looking at the signature of `register_job`
```
    def register_job(
        self,
        target: Callable,
        *args: Any,
        repeat_interval_s: Optional[PositiveFloat] = None,
    ) -> None:
```

You can see `target` and `*args`. These are provided by the user of this class. The handler logic and `*args` are not bound to any specific pattern. So I'd sat that from point of view of separation of concerns this is OK. I think it is better decoupled like this.

In this situation, it is just a coincidence that settings is the only parameter and that the `repeat_interval_s` is also contained inside settings.